lib/websocket.js

```diff
@@ -17,29 +17,120 @@ const OPCODE_SHORT = 0x81;
 const LEN_16_BIT = 126;
 const MAX_16_BIT = 65536;
 const LEN_64_BIT = 127;
+const MAX_64_BIT = 0x7fffffffffffffffn + 1n;
 
-const calcOffset = (frame, length) => {
-  if (length < LEN_16_BIT) return [2, 6];
-  if (length === LEN_16_BIT) return [4, 8];
-  return [10, 14];
+const DEFAULT_OPTIONS = {
+  server: true,
 };
 
-const parseFrame = (frame) => {
-  const length = frame[1] ^ 0x80;
-  const [maskOffset, dataOffset] = calcOffset(frame, length);
-  const mask = frame.subarray(maskOffset, maskOffset + MASK_LENGTH);
-  const data = frame.subarray(dataOffset);
-  return { mask, data };
-};
+const xor = (cond1, cond2) => (cond1 || cond2) && !(cond1 && cond2);
+
+class Frame {
```

Comment:
Typically in a constructor you just assign some props for instance. Make some configuration. But you don't do some complex logic of how inputs should be mapped correctly to Frame class props. This is not the responsibility of the constructor. So we can create `from` method that will take this responsibility. See in the example below. Also we should not handle encode or decode logic in constructor. Because when you create frame with string or buffer there is no difference. `new Frame('1234')` and `new Frame(buffer)`. Code becomes implicit and less readable. Instead you can do `Frame.from(data).decode()` and `Frame.from(data).decode()`. This makes clear what happens in this code. See in the example below. So, please use interface described in example below.

```
class Frame {
  #buffer = null

  constructor(buffer) {
    this.#buffer = buffer
  }


  encode() {
    ...
  }

  decode() {
    ...
  }


  static from(data) {
    if (Buffer.isBuffer(data)) {
      return new Frame(data)
    } 

    if (typeof data === 'string') {
      return new Frame(Buffer.from(data))
    }

    throw new Error('Unsupported')
  }
}

// USAGE

send(text) {
  const frame = Frame.from(text).encode()
  this.socket.write(frame.data);
}

receive(data) {
  console.log('data: ', data[0], data.length);
  if (data[0] !== OPCODE_SHORT) return;
  const frame = Frame.from(data).decode()
  const text = frame.toString();
  this.send(`Echo "${text}"`);
  console.log('Message:', text);
}


```

Reply:
Okay, with the factory, yeah, that's probably better. But creating a text frame of a protocol from a string and just saving the finished frame in a class field is not the same action. Then it is probably better to just accept a ready frame by constructor and create a protocol frame from input data by static method. Thanks.